### PR TITLE
Setup/IN: Amazon S3 or Cloudinary #41

### PR DESCRIPTION
**Pull request by luisreyesxv**
### What this PR does (required):
-set up AWS s3 bucket
- Created back-end route that takes takes image file and uploads it to S3 bucket
- created new field for User model to hold the avatar

### Screenshots / Videos (required):
Screenshot 2021-09-16 070701

### Any information needed to test this feature (required):
- need to re-run npm install for the new packages
- inside the slack chat, there are a few api variables that need to be added need to be added to the .env file. Look there for them. I can also provide them directly

### Any issues with the current functionality (optional):
-The ticket wanted to integrate a a way for the user to upload the profile picture in the front end, but there ticket #24 is also asking for this. So I to not deal with 2 tickets in one, I left the code to only be the backend route side.

### CODE REVIEW

File: server/controllers/image.js [hunk 1]
```diff
@@ -0,0 +1,30 @@
+const User = require("../models/User");
+const asyncHandler = require("express-async-handler");
+
+// @route PUT /avatar
+// @desc add or updates user's avatar image
+// @access Private
+exports.updateAvatar = asyncHandler(async (req, res, next) => {
+  const user = await User.findById(req.user.id);
+  const avatarURL = req.file.location
```
Comment (ezeilo-su) [inline server/controllers/image.js:9]: Would be good to validate this one too `file` and `location`

File: server/controllers/image.js [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+  if (!user) {
+    res.status(401);
```
Comment (ezeilo-su) [inline server/controllers/image.js:12]: :+1:

File: server/controllers/image.js [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+    throw new Error("Not authorized")
+  }
+  try {
+
+    user.editAvatar(avatarURL)
```
Comment (ezeilo-su) [inline server/controllers/image.js:17]: Validate `avatarURL` and return an error status if it's not supplied in the request.

File: server/models/User.js [hunk 4]
```diff
@@ -35,4 +41,13 @@ userSchema.pre("save", async function (next) {
   this.password = await bcrypt.hash(this.password, salt);
 });
 
+// instance method to takes a file and uploads it online and stores a new avatar link to document
+userSchema.methods.editAvatar = async function (url) {
+
+  this.avatar = url
+  this.save()
```
Comment (ezeilo-su) [inline server/models/User.js:48]: Mongoose `save` method is a `Promise`. So, it's good to `await` it, even though it's the last line of code here.

File: server/routes/image.js [hunk 5]
```diff
@@ -0,0 +1,14 @@
+const express = require('express');
+const router = express.Router();
+const { singleUpload } = require('../middleware/imageUpload');
+const protect = require('../middleware/auth');
+const {
+  updateAvatar
+} = require('../controllers/image');
+
+router.route('/avatar').put(protect, singleUpload, updateAvatar);
+// router.route('/showcase').put(protect, singleUpload);
```
Comment (ezeilo-su) [inline server/routes/image.js:10]: Remove unused code!

**Review by ezeilo-su — COMMENTED** [review]
Good work on implementing this one :star: 
Good PR message, screenshot, referencing the issue.

File: server/middleware/imageUpload.js [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+const multerS3 = require('multer-s3')
+const s3Connect = require("../utils/aws")
+
+const acceptedFileTypes = ["image/jpeg", "image/png"]
+
+const fileFilter = (req, file, cb) => {
+  // cb is for multer and its a callback function that gets called returned when multer finishes.
+  // per convos I found online, its better practice to pass null when everything is cool, and throw an error when something doesn't "match"
+  // what I want.
+  if (acceptedFileTypes.includes(file.mimetype)) {
```
Comment (rajivtitus) [inline server/middleware/imageUpload.js:11]: 👍🏽

**Comment by rajivtitus** [comment]
Good work with this! :)

Just address pending comments and then tag me in to get this approved.

**Review by rajivtitus — APPROVED** [review]
LGTM! Thanks for making requested changes.

**Pull request merged**